src/core/mempool.rs

```diff
@@ -1360,16 +1640,231 @@ impl MemPoolDB {
+        } else {
+            Ok(MemPoolSyncData::BloomFilter(self.get_txid_bloom_filter()?))
+        }
+    }
+
+    /// Get the next batch of transactions from our mempool that are *not* represented in the given
+    /// MemPoolSyncData.  Transactions are ordered lexicographically by randomized_txids.hashed_txid, since this allows us
+    /// to use the txid as a cursor while ensuring that each node returns txids in a deterministic random order
+    /// (so if some nodes are configured to return fewer than MAX_BLOOM_COUNTER_TXS transactions,
+    /// a requesting node will still have a good chance of getting something useful).
+    pub fn find_next_missing_transactions(
+        &self,
+        data: &MemPoolSyncData,
+        height: u64,
+        last_txid: &Txid,
+        max_txs: u64,
+        max_run: u64,
+    ) -> Result<Vec<StacksTransaction>, db_error> {
+        let mut ret = vec![];
+        let sql = "SELECT mempool.txid as txid, mempool.tx as tx \
+                   FROM mempool JOIN randomized_txids \
+                   ON mempool.txid = randomized_txids.txid \
+                   WHERE randomized_txids.hashed_txid > ?1 \
+                   AND mempool.height > ?2 \
+                   AND NOT EXISTS \
+                        (SELECT 1 FROM removed_txids WHERE txid = mempool.txid) \
+                   ORDER BY randomized_txids.hashed_txid ASC LIMIT ?3";
+
+        let args: &[&dyn ToSql] = &[
+            &last_txid,
+            &u64_to_sql(height.saturating_sub(BLOOM_COUNTER_DEPTH as u64))?,
+            &u64_to_sql(max_run)?,
+        ];
+
+        let mut tags_table = HashSet::new();
+        if let MemPoolSyncData::TxTags(_, ref tags) = data {
+            for tag in tags.iter() {
+                tags_table.insert(tag.clone());
+            }
+        }
+
+        let mut stmt = self.conn().prepare(sql)?;
+        let mut rows = stmt.query(args)?;
+        while let Some(row) = rows.next()? {
+            let txid = Txid::from_column(row, "txid")?;
+            test_debug!("Consider txid {}", &txid);
+            let contains = match data {
+                MemPoolSyncData::BloomFilter(ref bf) => bf.contains_raw(&txid.0),
+                MemPoolSyncData::TxTags(ref seed, ..) => {
+                    tags_table.contains(&TxTag::from_seed_and_txid(seed, &txid))
+                }
+            };
+            if contains {
+                // remote peer already has this one
+                continue;
+            }
+
+            let tx_bytes: Vec<u8> = row.get_unwrap("tx");
+            let tx = StacksTransaction::consensus_deserialize(&mut &tx_bytes[..])
+                .map_err(|_e| db_error::ParseError)?;
+
+            test_debug!("Returning txid {}", &txid);
+            ret.push(tx);
+            if (ret.len() as u64) >= max_txs {
+                break;
+            }
+        }
+
+        Ok(ret)
+    }
+
+    /// Stream transaction data
+    pub fn stream_txs<W: Write>(
```

Comment:
Can you make this more descriptive? Is this method supposed to (1) query the mempool using the MemPoolSyncData query contained in the TxStreamData, (2) buffer the resulting transactions inside of the TxStreamData struct, and then (3) write from the TxStreamData struct's buffer into `fd`? That seems like something that could be called `handle_mempool_transaction_query` or something similar.

I also wonder about the structure of this streaming -- why use the TxStreamData to buffer data before writing to the fd? First, why buffer at all instead of just writing from the `find_next_missing_transactions` result into the fd. Second, why use the TxStreamData struct for the buffer, instead of a local buffer?

Reply:
So, this method is supposed to feed at most `count` bytes into `fd` from a stream of transactions loaded from the mempool that match a `MemPoolSyncData` query (owned by the given `TxStreamData`).  The `TxStreamData` is just a cursor -- it's all the state the HTTP server needs to remember in-between attempts to fill the underlying socket buffer with more transaction data as it becomes writeable.

While `fd` is just something that implements `Write`, in practice `fd` is a `PipeWrite` instance, which permits the networking code to buffer up messages for transmission without duplicating a lot of buffer book-keeping work.  The socket server consumes data from the corresponding `PipeRead` and writes *that* to the actual socket.  The reason `Pipe`, `PipeRead`, and `PipeWrite` exist as this layer of indirection is because the socket server doesn't know how many bytes are going to be available in the socket buffer until after it becomes writeable.  By passing a `PipeWrite` to `consensus_serialize()`, we ensure that `consensus_serialize()` always runs to completion without blocking, and the socket server takes care of reading the data out from its `PipeRead` in order to fill up socket buffer bytes as more space becomes available.  The various streaming methods, including this one, take a `count` argument to instruct the implementation to only push at most `count` bytes into the `PipeWrite` in order to keep memory usage under control.  The alternative is far messier -- we'd either have to write all our code assuming that `consensus_serialize()` could block for a long, long time, or implement `consensus_serialize()` to be "restartable" at a given byte offset.

Now, the reason we have a `tx_buf` at all in the `TxStreamData` is because we need the transaction being queried to stick around across multiple calls to `stream_txs` (which can be interleaved with mempool garbage collection), which necessitates copying it out of the mempool and into this cursor via `find_next_missing_transactions()`.  We could just as easily stored the `StacksTransaction` in the `TxStreamData`, but that would necessitate re-serializing the transaction each time the socket server asked for more bytes.  Using `tx_buf` and `tx_buf_ptr` is an optimization that ensures we only load and serialize the transaction once.

Reply:
Also, the reason it's called `stream_txs` is that it follows the naming scheme of `stream_headers`, `stream_block`, `stream_microblocks_confirmed`, and `stream_microblocks_unconfirmed`.  In all cases, these method write at most `count` bytes to a given `Write` called `fd` (really a `PipeWrite`), and use state within their respective `StreamCursor` variants to keep track of what to write on the subsequent call.

Reply:
Okay, I see why `query.tx_buff` is used in this case. I still think this function needs to be named something else and have an expanded rustdoc: `stream_txs` tells us relatively little about what this function is doing -- a function named like that implies that it's just blindly streaming transactions from one buffer into another, but this function is doing substantially more work than that -- maintaining request state, querying the database, etc., that's much closer to `handle_mempool_transaction_query`. (I'm not sure the other `stream_x` functions are well-named, either, but they're not directly related to this PR).